plugin/src/main/java/org/opensearch/ml/action/models/UpdateModelTransportAction.java

```diff
@@ -338,56 +414,132 @@ private void updateRequestConstructor(
+                            e
+                        );
+                    wrappedListener.onFailure(e);
+                }));
+            } else {
+                client.update(updateModelGroupRequest, ActionListener.wrap(r -> {
+                    client.update(updateRequest, getUpdateResponseListener(modelId, wrappedListener));
+                }, e -> {
+                    log
+                        .error(
+                            "Failed to register ML model with model ID {} to the new model group with model group ID {}",
+                            modelId,
+                            newModelGroupId,
+                            e
+                        );
+                    wrappedListener.onFailure(e);
+                }));
+            }
         } catch (IOException e) {
             log.error("Failed to build update request.");
-            actionListener.onFailure(e);
+            wrappedListener.onFailure(e);
         }
     }
 
-    private ActionListener<UpdateResponse> getUpdateResponseListener(
+    private ActionListener<UpdateResponse> getUpdateResponseListenerWithUpdateModelCache(
         String modelId,
-        ActionListener<UpdateResponse> actionListener,
-        ThreadContext.StoredContext context
+        ActionListener<UpdateResponse> wrappedListener,
+        MLUpdateModelCacheNodesRequest mlUpdateModelCacheNodesRequest
     ) {
-        return ActionListener.runBefore(ActionListener.wrap(updateResponse -> {
-            if (updateResponse != null && updateResponse.getResult() != DocWriteResponse.Result.UPDATED) {
-                log.info("Model id:{} failed update", modelId);
-                actionListener.onResponse(updateResponse);
-                return;
+        return ActionListener.wrap(updateResponse -> {
+            if (updateResponse != null && updateResponse.getResult() == DocWriteResponse.Result.UPDATED) {
+                client.execute(MLUpdateModelCacheAction.INSTANCE, mlUpdateModelCacheNodesRequest, ActionListener.wrap(r -> {
+                    if (isUpdateModelCacheSuccessOnAllNodes(modelId, r)) {
+                        log.info("Successfully updated ML model cache with model ID {}", modelId);
+                        wrappedListener.onResponse(updateResponse);
+                    } else {
+                        String[] nodeIds = getUpdateModelCacheFailedNodesList(modelId, r);
+                        log
+                            .error(
+                                "Successfully update ML model index with model ID {} but update model cache was failed on following nodes {}",
+                                modelId,
+                                Arrays.toString(nodeIds)
+                            );
+                        wrappedListener
+                            .onFailure(
+                                new RuntimeException(
+                                    "Successfully update ML model index with model ID"
```

Comment:
Add suggested action for user: Please retry

Reply:
Added